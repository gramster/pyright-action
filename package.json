{
    "name": "pyright-action",
    "version": "1.3.0",
    "private": true,
    "description": "GitHub Action for pyright",
    "main": "dist/index.js",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/jakebailey/pyright-action.git"
    },
    "author": "Jake Bailey",
    "license": "MIT",
    "engines": {
        "node": ">=16"
    },
    "dependencies": {
        "@actions/core": "^1.8.2",
        "@actions/http-client": "^2.0.1",
        "@actions/tool-cache": "^2.0.1",
        "@badrap/valita": "^0.1.3",
        "semver": "^7.3.7",
        "string-argv": "^0.3.1"
    },
    "devDependencies": {
        "@tsconfig/node16": "^1.0.2",
        "@types/clone": "^2.1.1",
        "@types/jest": "^27.5.1",
        "@types/node": "^16.11.36",
        "@types/semver": "^7.3.9",
        "@types/semver-utils": "^1.1.1",
        "@typescript-eslint/eslint-plugin": "^5.24.0",
        "@typescript-eslint/parser": "^5.24.0",
        "clone": "^2.1.2",
        "esbuild": "^0.14.39",
        "eslint": "^8.15.0",
        "eslint-config-prettier": "^8.5.0",
        "eslint-plugin-simple-import-sort": "^7.0.0",
        "husky": "^8.0.1",
        "jest": "^28.1.0",
        "jest-serializer-path": "^0.1.15",
        "npm-check-updates": "^13.0.0",
        "prettier": "2.6.2",
        "release-it": "^15.0.0",
        "rimraf": "^3.0.2",
        "ts-jest": "^28.0.2",
        "typescript": "~4.6.4"
    },
    "scripts": {
        "postinstall": "husky install",
        "build": "esbuild src/index.ts --bundle --outfile=dist/index.js --platform=node --target=node16",
        "update:all": "npm-check-updates -u && yarn",
        "update:all:transitive": "npm-check-updates -u && rimraf yarn.lock && yarn",
        "release": "release-it",
        "test": "jest"
    },
    "release-it": {
        "git": {
            "commitMessage": "Release v${version}",
            "tagName": "v${version}"
        },
        "npm": {
            "publish": false
        },
        "github": {
            "release": true,
            "releaseName": "v${version}",
            "web": true
        },
        "hooks": {
            "before:init": "yarn test",
            "after:git:release": "git tag -d v1 && git push origin :refs/tags/v1 && git tag v1 && git push origin v1"
        }
    },
    "packageManager": "yarn@3.2.1",
    "volta": {
        "node": "16.15.0"
    }
}
