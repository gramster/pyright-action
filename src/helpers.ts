import * as core from '@actions/core';
import * as httpClient from '@actions/http-client';
import * as tc from '@actions/tool-cache';
import * as path from 'path';
import SemVer from 'semver/classes/semver';
import stringArgv from 'string-argv';

import { version as actionVersion } from '../package.json';
import { NpmRegistryResponse, parseNpmRegistryResponse } from './schema';

export function getActionVersion() {
    return actionVersion;
}

export function getNodeInfo() {
    return {
        version: process.version,
        execPath: process.execPath,
    };
}

export async function getArgs() {
    const pyrightInfo = await getPyrightInfo();
    const pyrightPath = await downloadPyright(pyrightInfo);

    const args = [path.join(pyrightPath, 'package', 'index.js')];

    const workingDirectory = core.getInput('working-directory');

    const noComments = getBooleanInput('no-comments', false);
    if (!noComments) {
        args.push('--outputjson');
    }

    const treatPartialAsWarning = getBooleanInput('warn-partial', false);
    const pythonPlatform = core.getInput('python-platform');
    if (pythonPlatform) {
        args.push('--pythonplatform');
        args.push(pythonPlatform);
    }

    const pythonVersion = core.getInput('python-version');
    if (pythonVersion) {
        args.push('--pythonversion');
        args.push(pythonVersion);
    }

    const typeshedPath = core.getInput('typeshed-path');
    if (typeshedPath) {
        args.push('--typeshed-path');
        args.push(typeshedPath);
    }

    const venvPath = core.getInput('venv-path');
    if (venvPath) {
        args.push('--venv-path');
        args.push(venvPath);
    }

    const project = core.getInput('project');
    if (project) {
        args.push('--project');
        args.push(project);
    }

    const lib = getBooleanInput('lib', false);
    if (lib) {
        args.push('--lib');
    }

    const warnings = getBooleanInput('warnings', false);
    if (warnings) {
        args.push('--warnings');
    }

    const verifyTypes = core.getInput('verify-types');
    if (verifyTypes) {
        args.push('--verifytypes');
        args.push(verifyTypes);
    }

    const extraArgs = core.getInput('extra-args');
    if (extraArgs) {
        args.push(...stringArgv(extraArgs));
    }

    return {
        workingDirectory,
        noComments,
        treatPartialAsWarning,
        pyrightVersion: pyrightInfo.version,
        args,
    };
}

function getBooleanInput(name: string, defaultValue: boolean): boolean {
    const input = core.getInput(name);
    if (!input) {
        return defaultValue;
    }
    return input.toUpperCase() === 'TRUE';
}

const pyrightToolName = 'pyright';

async function downloadPyright(info: NpmRegistryResponse): Promise<string> {
    // Note: this only works because the pyright package doesn't have any
    // dependencies. If this ever changes, we'll have to actually install it.
    const found = tc.find(pyrightToolName, info.version);
    if (found) {
        return found;
    }

    const tarballPath = await tc.downloadTool(info.dist.tarball);
    const extractedPath = await tc.extractTar(tarballPath);
    return await tc.cacheDir(extractedPath, pyrightToolName, info.version);
}

async function getPyrightInfo(): Promise<NpmRegistryResponse> {
    const version = getPyrightVersion();
    const client = new httpClient.HttpClient();
    const resp = await client.get(`https://registry.npmjs.org/pyright/${version}`);
    const body = await resp.readBody();
    if (resp.message.statusCode !== httpClient.HttpCodes.OK) {
        throw new Error(body);
    }
    return parseNpmRegistryResponse(JSON.parse(body));
}

function getPyrightVersion() {
    const versionSpec = core.getInput('version');
    if (versionSpec) {
        return new SemVer(versionSpec).format();
    }
    return 'latest';
}
