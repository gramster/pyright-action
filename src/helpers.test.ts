import * as core from '@actions/core';
import * as httpClient from '@actions/http-client';
import * as tc from '@actions/tool-cache';
import type { IncomingMessage } from 'http';
import serializer from 'jest-serializer-path';
import * as os from 'os';
import * as path from 'path';

expect.addSnapshotSerializer(serializer);

jest.mock('@actions/core');
const mockedCore = jest.mocked(core);
jest.mock('@actions/http-client');
const mockedHttpClient = jest.mocked(httpClient, true);
jest.mock('@actions/tool-cache');
const mockedTc = jest.mocked(tc);

import { version as actionVersion } from '../package.json';
import { getActionVersion, getArgs, getNodeInfo } from './helpers';
import { NpmRegistryResponse } from './schema';

const fakeRoot = path.join(os.tmpdir(), 'rootDir');

beforeEach(() => {
    jest.clearAllMocks();
    mockedTc.find.mockReturnValue('');
    mockedTc.cacheDir.mockImplementation(async (dir) => path.join(fakeRoot, 'cached', path.relative(fakeRoot, dir)));
});

afterEach(() => {
    expect(mockedCore.getInput.mock.calls).toMatchSnapshot('core.getInput');
    expect(mockedTc.downloadTool.mock.calls).toMatchSnapshot('tc.downloadTool');
    expect(mockedTc.extractTar.mock.calls).toMatchSnapshot('tc.extractTar');
    expect(mockedTc.find.mock.calls).toMatchSnapshot('tc.find');
    expect(mockedTc.cacheDir.mock.calls).toMatchSnapshot('tc.cacheDir');
});

describe('getArgs', () => {
    test('bad version', async () => {
        mockedCore.getInput.mockImplementation((name, options) => {
            expect(options).toBeUndefined();
            switch (name) {
                case 'version':
                    return 'this is not a semver';
                default:
                    return '';
            }
        });

        await expect(getArgs()).rejects.toThrowError('not a semver');
    });

    describe('valid version', () => {
        const npmResponse: NpmRegistryResponse = {
            version: '1.1.240',
            dist: {
                tarball: 'https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz',
            },
        };
        const tarballPath = path.join(fakeRoot, 'pyright.tar.gz');
        const extractedPath = path.join(fakeRoot, 'pyright');

        const inputs = new Map<string, string>();

        beforeEach(() => {
            inputs.clear();

            mockedCore.getInput.mockImplementation((name, options) => {
                expect(options).toBeUndefined();
                return inputs.get(name) ?? '';
            });

            mockedHttpClient.HttpClient.prototype.get.mockImplementation(async (url) => {
                switch (url) {
                    case `https://registry.npmjs.org/pyright/${npmResponse.version}`:
                    case 'https://registry.npmjs.org/pyright/latest':
                        return {
                            message: {
                                statusCode: 200,
                            } as IncomingMessage,
                            readBody: async () => JSON.stringify(npmResponse),
                        };
                    case `https://registry.npmjs.org/pyright/1.1.404`:
                        return {
                            message: {
                                statusCode: 404,
                            } as IncomingMessage,
                            readBody: async () => JSON.stringify('version not found: 1.1.404'),
                        };
                    default:
                        throw new Error(`unknown URL ${url}`);
                }
            });

            mockedTc.downloadTool.mockResolvedValue(tarballPath);
            mockedTc.extractTar.mockResolvedValue(extractedPath);
        });

        test('many options', async () => {
            inputs.set('working-directory', '/path/to/project');
            inputs.set('python-platform', 'Linux');
            inputs.set('python-version', '3.9');
            inputs.set('typeshed-path', '/path/to/typeshed');
            inputs.set('venv-path', '/path/to-venv');
            inputs.set('project', '/path/to/pyrightconfig.json');
            inputs.set('lib', 'True');
            inputs.set('warnings', 'TrUe');
            inputs.set('extra-args', '--foo --bar --baz');

            const result = await getArgs();
            expect(result).toMatchSnapshot('result');

            expect(mockedTc.downloadTool).toBeCalledWith(npmResponse.dist.tarball);
            expect(mockedTc.extractTar).toBeCalledWith(tarballPath);
        });

        test('no comments', async () => {
            inputs.set('no-comments', 'true');

            const result = await getArgs();
            expect(result).toMatchSnapshot('result');
        });

        test('verifytypes', async () => {
            inputs.set('verify-types', 'some.package');

            const result = await getArgs();
            expect(result).toMatchSnapshot('result');
        });

        test('version not found', async () => {
            inputs.set('version', '1.1.404');
            await expect(getArgs()).rejects.toThrowError('version not found: 1.1.404');
        });

        test('cached', async () => {
            mockedTc.find.mockReturnValue(path.join(fakeRoot, 'cached', 'pyright'));

            const result = await getArgs();
            expect(result).toMatchSnapshot('result');
            expect(mockedTc.cacheDir).toBeCalledTimes(0);
        });
    });
});

test('getNodeInfo', () => {
    const info = getNodeInfo();
    expect(info).toEqual({
        version: process.version,
        execPath: process.execPath,
    });
});

test('getActionVersion', () => {
    const version = getActionVersion();
    expect(version).toEqual(actionVersion);
});
