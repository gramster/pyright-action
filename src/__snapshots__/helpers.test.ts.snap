// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getActionVersion: core.getInput 1`] = `Array []`;

exports[`getActionVersion: tc.cacheDir 1`] = `Array []`;

exports[`getActionVersion: tc.downloadTool 1`] = `Array []`;

exports[`getActionVersion: tc.extractTar 1`] = `Array []`;

exports[`getActionVersion: tc.find 1`] = `Array []`;

exports[`getArgs bad version: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
]
`;

exports[`getArgs bad version: tc.cacheDir 1`] = `Array []`;

exports[`getArgs bad version: tc.downloadTool 1`] = `Array []`;

exports[`getArgs bad version: tc.extractTar 1`] = `Array []`;

exports[`getArgs bad version: tc.find 1`] = `Array []`;

exports[`getArgs valid version cached: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
  Array [
    "working-directory",
  ],
  Array [
    "no-comments",
  ],
  Array [
    "python-platform",
  ],
  Array [
    "python-version",
  ],
  Array [
    "typeshed-path",
  ],
  Array [
    "venv-path",
  ],
  Array [
    "project",
  ],
  Array [
    "lib",
  ],
  Array [
    "warnings",
  ],
  Array [
    "verify-types",
  ],
  Array [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version cached: result 1`] = `
Object {
  "args": Array [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
    "--outputjson",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version cached: tc.cacheDir 1`] = `Array []`;

exports[`getArgs valid version cached: tc.downloadTool 1`] = `Array []`;

exports[`getArgs valid version cached: tc.extractTar 1`] = `Array []`;

exports[`getArgs valid version cached: tc.find 1`] = `
Array [
  Array [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version many options: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
  Array [
    "working-directory",
  ],
  Array [
    "no-comments",
  ],
  Array [
    "python-platform",
  ],
  Array [
    "python-version",
  ],
  Array [
    "typeshed-path",
  ],
  Array [
    "venv-path",
  ],
  Array [
    "project",
  ],
  Array [
    "lib",
  ],
  Array [
    "warnings",
  ],
  Array [
    "verify-types",
  ],
  Array [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version many options: result 1`] = `
Object {
  "args": Array [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
    "--outputjson",
    "--pythonplatform",
    "Linux",
    "--pythonversion",
    "3.9",
    "--typeshed-path",
    "/path/to/typeshed",
    "--venv-path",
    "/path/to-venv",
    "--project",
    "/path/to/pyrightconfig.json",
    "--lib",
    "--warnings",
    "--foo",
    "--bar",
    "--baz",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "/path/to/project",
}
`;

exports[`getArgs valid version many options: tc.cacheDir 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version many options: tc.downloadTool 1`] = `
Array [
  Array [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version many options: tc.extractTar 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version many options: tc.find 1`] = `
Array [
  Array [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version no comments: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
  Array [
    "working-directory",
  ],
  Array [
    "no-comments",
  ],
  Array [
    "python-platform",
  ],
  Array [
    "python-version",
  ],
  Array [
    "typeshed-path",
  ],
  Array [
    "venv-path",
  ],
  Array [
    "project",
  ],
  Array [
    "lib",
  ],
  Array [
    "warnings",
  ],
  Array [
    "verify-types",
  ],
  Array [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version no comments: result 1`] = `
Object {
  "args": Array [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
  ],
  "noComments": true,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version no comments: tc.cacheDir 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version no comments: tc.downloadTool 1`] = `
Array [
  Array [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version no comments: tc.extractTar 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version no comments: tc.find 1`] = `
Array [
  Array [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version verifytypes: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
  Array [
    "working-directory",
  ],
  Array [
    "no-comments",
  ],
  Array [
    "python-platform",
  ],
  Array [
    "python-version",
  ],
  Array [
    "typeshed-path",
  ],
  Array [
    "venv-path",
  ],
  Array [
    "project",
  ],
  Array [
    "lib",
  ],
  Array [
    "warnings",
  ],
  Array [
    "verify-types",
  ],
  Array [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version verifytypes: result 1`] = `
Object {
  "args": Array [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
    "--outputjson",
    "--verifytypes",
    "some.package",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version verifytypes: tc.cacheDir 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.downloadTool 1`] = `
Array [
  Array [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.extractTar 1`] = `
Array [
  Array [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.find 1`] = `
Array [
  Array [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version version not found: core.getInput 1`] = `
Array [
  Array [
    "version",
  ],
]
`;

exports[`getArgs valid version version not found: tc.cacheDir 1`] = `Array []`;

exports[`getArgs valid version version not found: tc.downloadTool 1`] = `Array []`;

exports[`getArgs valid version version not found: tc.extractTar 1`] = `Array []`;

exports[`getArgs valid version version not found: tc.find 1`] = `Array []`;

exports[`getNodeInfo: core.getInput 1`] = `Array []`;

exports[`getNodeInfo: tc.cacheDir 1`] = `Array []`;

exports[`getNodeInfo: tc.downloadTool 1`] = `Array []`;

exports[`getNodeInfo: tc.extractTar 1`] = `Array []`;

exports[`getNodeInfo: tc.find 1`] = `Array []`;
